libbeat/publisher/pipeline/output.go

```diff
@@ -18,115 +18,130 @@
 package pipeline
 
 import (
-	"github.com/elastic/beats/v7/libbeat/common/atomic"
 	"github.com/elastic/beats/v7/libbeat/logp"
 	"github.com/elastic/beats/v7/libbeat/outputs"
 )
 
-// clientWorker manages output client of type outputs.Client, not supporting reconnect.
-type clientWorker struct {
+type worker struct {
 	observer outputObserver
 	qu       workQueue
-	client   outputs.Client
-	closed   atomic.Bool
+	done     chan struct{}
+}
+
+// clientWorker manages output client of type outputs.Client, not supporting reconnect.
+type clientWorker struct {
+	worker
+	client outputs.Client
 }
 
 // netClientWorker manages reconnectable output clients of type outputs.NetworkClient.
 type netClientWorker struct {
-	observer outputObserver
-	qu       workQueue
-	client   outputs.NetworkClient
-	closed   atomic.Bool
+	worker
+	client outputs.NetworkClient
 
 	batchSize  int
 	batchSizer func() int
 	logger     *logp.Logger
 }
 
 func makeClientWorker(observer outputObserver, qu workQueue, client outputs.Client) outputWorker {
+	w := worker{
+		observer: observer,
+		qu:       qu,
+		done:     make(chan struct{}),
+	}
+
+	var c interface {
+		outputWorker
+		run()
+	}
+
 	if nc, ok := client.(outputs.NetworkClient); ok {
-		c := &netClientWorker{
-			observer: observer,
-			qu:       qu,
-			client:   nc,
-			logger:   logp.NewLogger("publisher_pipeline_output"),
+		c = &netClientWorker{
+			worker: w,
+			client: nc,
+			logger: logp.NewLogger("publisher_pipeline_output"),
 		}
-		go c.run()
-		return c
+	} else {
+		c = &clientWorker{worker: w, client: client}
 	}
-	c := &clientWorker{observer: observer, qu: qu, client: client}
+
 	go c.run()
 	return c
 }
 
 func (w *clientWorker) Close() error {
-	w.closed.Store(true)
+	close(w.worker.done)
 	return w.client.Close()
 }
 
 func (w *clientWorker) run() {
-	for !w.closed.Load() {
-		for batch := range w.qu {
-			w.observer.outBatchSend(len(batch.events))
+	for {
+		// We wait for either the worker to be closed or for there to be a batch of
+		// events to publish.
+		select {
 
+		case <-w.done:
+			return
```

Comment:
Not sure, just a hunch. But please check the state of batches and the workqueue in case we return early. Is the workqueue reused or do we create a new workqueue for the new set of outputs? In the former case we should be fine. But the original code did wait for the work queue to be closed. This way we've been sure to not have any batches in flight. By returning early there is a chance of having batches still in the workqueue. Is the queue drained somewhere else? At some point (maybe a defer here) we need a draining loop like this:

```
// cancel -> return to retryer -> rescheduled batches with new queue
for batch := range w.qu {
  batch.Cancelled()
}
```

or like this:

```
// transfer pending batches to new workqueue
for batch := range w.qu {
  newQueue <- batch
}
```

Reply:
Okay, I checked this and it looks like a **new workqueue** is created each time we set a new group of outputs on the controller:

https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/libbeat/publisher/pipeline/controller.go#L103

The retryer is then given this new workqueue (assigned to `retryer.out`). So canceling the batch from the old worker so it's returned to the retry queue (`retryer.in`) makes sense here.

Reply:
Implemented in 400a5d4d4.